src/simsopt/mhd/spec.py

```diff
@@ -215,9 +739,14 @@ def init(self, filename: str):
         spec.preset()
         logger.debug("Done with init")
 
-    def run(self):
+    def run(self, update_guess: bool = True):
```

Comment:
To what extent does the default `update_guess=True` confuse finite differencing and make optimization depend on the number of MPI processes? If the initial guesses for the axis and internal surfaces are not the same for every function evaluation, then the objective function depends slightly on the history of previous objective function evaluations on that process. That history in turn depends on how many MPI processes are used. For finite difference derivatives, where tiny differences in the objective function are particularly important, this could be an issue.

Reply:
Good point - I haven't thought about this.

My first approach would be to set `update_guess=False` whenever a worker calls `spec.run()` to evaluate a finite-difference derivative, and only call with `update_guess=True` when a step is taken by the optimizer. I guess this means we would have to modify `least_square_mpi_solve._f_proc0` and `least_square_mpi_solve._constrained_mpi_workers_task`?

I don't think this is ideal - I would prefer to only modify `spec.py`. Any idea?

Reply:
> When evaluating finite differences, only small deviations dx of each degree of freedom x are considered around x_0. The equilibrium obtained by SPEC at x_0+dx is thus very close to the one at x_0, and can also be used as an initial guess. The way I understand it is that the objective function depends on SPEC equilibrium, but _does not_ depend on SPEC initial guess. 

Ideally, if the equations are solved in spec to very tight tolerance, the objective should not depend on the initial guess. But for finite tolerances, the initial guess may affect the result a bit, and it only takes a small change to the objective to cause an O(1) change to a derivative computed by finite differencing. As an indication that we might want to worry about this, consider the issue discussed in this PR of the residue changing by 0.5% when the coordinate axis shape changed. Also note that when there are multiple worker groups, when an MPI process in group > 0 starts evaluating the objective at x_j + dx for a Jacobian, the initial condition will come from a run from the previous Jacobian evaluation, at x_{j-1}, which may not be super close to x_j.

It's fine with me to have this default `update_guess = True`, since it's working for you. Just wanted to flag this in case it affects your calculations.

Reply:
You are right, we do not necessarily solve SPEC equations to machine precision, which might cause issues when evaluating finite differences. As I see it, we have two choices:

(1) Leave it as it is, with the risk that the different initial guesses might affect the Jacobian evaluation. 

(2) Enforce the initial guess to be the same for each finite-difference evaluation. The only (simple) way to do it is to modify the calls to the objective function from the solver and give an additional argument that contains the information on whether or not this call is for a finite-difference evaluation.

For example, [in finite_difference.py](https://github.com/hiddenSymmetries/simsopt/blob/c3f89c83a3013a2ea9ce7cbfaa66f6bbed0d3caf/src/simsopt/_core/finite_difference.py#LL256C42-L256C42),
```python
options = dict('finite_diff'=True)
out = np.asarray(self.fn( options ))
```
and then, in [spec.py](https://github.com/hiddenSymmetries/simsopt/blob/c3f89c83a3013a2ea9ce7cbfaa66f6bbed0d3caf/src/simsopt/mhd/spec.py#L218),
```python
def run(self, options):
    if options['finite_diff']:
        update_guess=False
    else:
        update_guess=True
```

The downside is that all objective functions must accept an `options` argument... which might require some changes in different files.

While option (2) is probably better, I think leaving it for another PR is best. What do you think?

Reply:
I agree that this can be addressed later in a separate pull request.